### ISSUE TITLE
Revisit softmax

### ISSUE THREAD

**Issue by adam-r-kowalski** [adam-r-kowalski/compute_graph#51]
The forward pass of `softmax` gives the same answer in my library, as well as tensorflow and pytorch. However, the backward pass gives wildly different gradient values for all three libraries. More research is needed to find out why.
